aws-codeguruprofiler-profilinggroup/src/main/java/software/amazon/codeguruprofiler/profilinggroup/UpdateHandler.java

```diff
@@ -0,0 +1,119 @@
+import software.amazon.awssdk.services.codeguruprofiler.model.PutPermissionRequest;
+import software.amazon.awssdk.services.codeguruprofiler.model.RemovePermissionRequest;
+import software.amazon.awssdk.services.codeguruprofiler.model.ResourceNotFoundException;
+import software.amazon.awssdk.services.codeguruprofiler.model.ThrottlingException;
+import software.amazon.awssdk.services.codeguruprofiler.model.ValidationException;
+import software.amazon.cloudformation.exceptions.CfnAlreadyExistsException;
+import software.amazon.cloudformation.exceptions.CfnInvalidRequestException;
+import software.amazon.cloudformation.exceptions.CfnNotFoundException;
+import software.amazon.cloudformation.exceptions.CfnServiceInternalErrorException;
+import software.amazon.cloudformation.exceptions.CfnThrottlingException;
+import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
+import software.amazon.cloudformation.proxy.Logger;
+import software.amazon.cloudformation.proxy.ProgressEvent;
+import software.amazon.cloudformation.proxy.ResourceHandlerRequest;
+
+import java.util.List;
+import java.util.Optional;
+
+public class UpdateHandler extends BaseHandler<CallbackContext> {
+
+    private final CodeGuruProfilerClient profilerClient = CodeGuruProfilerClientBuilder.create();
+
+    @Override
+    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
+        AmazonWebServicesClientProxy proxy,
+        ResourceHandlerRequest<ResourceModel> request,
+        CallbackContext callbackContext,
+        Logger logger) {
+        final ResourceModel model = request.getDesiredResourceState();
+
+        final String awsAccountId = request.getAwsAccountId();
+        final String profilingGroupName = model.getProfilingGroupName();
+
+        Optional<List<String>> principals = principalsForAgentPermissionsFrom(model);
+
+        try {
+            GetPolicyResponse getPolicyResponse = getExistingPolicy(proxy, profilingGroupName);
+
+            if (principals.isPresent()) {
+                putAgentPermissions(proxy, profilingGroupName, principals.get(), getPolicyResponse.revisionId());
+                logger.log(
+                    String.format("Policy for [%s] for accountId [%s] has been successfully updated! actionGroup: %s, principals: %s",
+                        profilingGroupName,
+                        awsAccountId,
+                        ActionGroup.AGENT_PERMISSIONS,
+                        principals.get())
+                );
+            } else if (getPolicyResponse.policy() != null) {
+                removeAgentPermission(proxy, profilingGroupName, getPolicyResponse.revisionId());
+                logger.log(String.format("Policy for [%s] for accountId [%s] has been successfully removed!", profilingGroupName, awsAccountId));
+            }
+
+            logger.log(String.format("%s [%s] for accountId [%s] has been successfully updated!", ResourceModel.TYPE_NAME, model.getProfilingGroupName(), awsAccountId));
```

Comment:
To be honest, if there is no change in the ProfilingGroup resource, this updateHandler would not be called at all. Therefore, we would always log what has been updated before this updated log to be called if UpdateHandler is invoked. 

Also, I like this updated statement outside the branch because it is the style we have been following and it is clear that we reached the end of the handler without issue.